pkg/controller/extension/controlplane/actuator.go

```diff
@@ -69,107 +62,81 @@ func NewActuator(mgr manager.Manager, config config.Configuration, extensionClas
 		scheme:            mgr.GetScheme(),
 		serviceConfig:     config,
 		extensionClasses:  extensionClasses,
-		certConfigDecoder: newCertConfigDecoder(mgr),
+		certConfigDecoder: shared.NewCertConfigDecoder(mgr),
 	}
 }
 
 type actuator struct {
-	certConfigDecoder
-	client           client.Client
-	config           *rest.Config
-	scheme           *runtime.Scheme
-	decoder          runtime.Decoder
-	extensionClasses []extensionsv1alpha1.ExtensionClass
+	certConfigDecoder shared.CertConfigDecoder
+	client            client.Client
+	config            *rest.Config
+	scheme            *runtime.Scheme
+	decoder           runtime.Decoder
+	extensionClasses  []extensionsv1alpha1.ExtensionClass
 
 	serviceConfig config.Configuration
 }
 
 // Reconcile the Extension resource.
 func (a *actuator) Reconcile(ctx context.Context, log logr.Logger, ex *extensionsv1alpha1.Extension) error {
-	var (
-		namespace         = ex.GetNamespace()
-		isShootDeployment = isShootDeployment(ex)
-		cluster           *extensions.Cluster
-		err               error
-	)
-
-	if isShootDeployment {
-		cluster, err = controller.GetCluster(ctx, a.client, namespace)
-		if err != nil {
-			return err
-		}
-	}
+	namespace := ex.GetNamespace()
 
-	certConfig, err := a.decodeAndValidateProviderConfig(ex, cluster)
+	certConfig, err := a.certConfigDecoder.DecodeAndValidateProviderConfig(ex, nil)
 	if err != nil {
 		return err
 	}
 
-	values, err := a.createValues(ctx, log, certConfig, cluster, namespace, ex)
+	values, err := a.createValues(certConfig, namespace, ex)
 	if err != nil {
 		return err
 	}
 
-	if isShootDeployment {
-		if !controller.IsHibernated(cluster) {
-			if err := a.createShootResourcesForShoot(ctx, log, *values); err != nil {
+	if err := a.createResourcesForGardenOrSeed(ctx, log, *values); err != nil {
+		return err
+	}
+
+	generateControlPlaneCertificate := ptr.Deref(certConfig.GenerateControlPlaneCertificate, false)
+	if isGardenDeployment(ex) {
+		handler := newGardenCert(a.client, log)
+		if generateControlPlaneCertificate {
+			if err := handler.reconcile(ctx, ex); err != nil {
+				return err
+			}
+		} else {
+			if err := handler.delete(ctx, ex); err != nil {
 				return err
 			}
-		}
-		if err := a.createSeedResourcesForShoot(ctx, log, *values); err != nil {
-			return err
 		}
 	} else {
-		if err := a.createResourcesForGardenOrSeed(ctx, log, *values); err != nil {
-			return err
-		}
+		handler := newControlPlaneCert(a.client, log)
+		if generateControlPlaneCertificate {
+			seedName := os.Getenv(EnvSeedName)
+			gardenClient, err := a.createGardenClient()
```

Comment:
This client could be created only once when initializing the the actuator, to reduce API discovery calls.